### PR TITLE
Dumper for token associations (tokenrels)

### LINKED ISSUE #9694: Differential testing: Dumper for token associations
Add a dumper for token associations

### PR DESCRIPTION
**Pull request by david-bakin-sl**
**Description**:

Dumps all token associations in a signed state file as CSV (actually, tab separated), deterministically.

**Related issue(s)**:

Fixes #9694.

**Example**:

`./services-cli.sh signed-state --file=states/2023-10-29T00:01:59Z/151975134/SignedState.swh  dump associations --associations=associations.txt --summary`

```
# ===  4805300 token associations (4805300 the hard way), 1467622 accounts with associations,   32997 token types associated with accounts
#   === token associations per account histogram ===
#         <=1:  1268337
#        <=10:   198265
#       <=100:     1012
#      <=1000:        8
account;token;balance;frozen;kycGranted;automaticAssociation;prev;next
200;2035606;1;T;T;;0;0
201;2035606;1;T;T;;0;0
202;2035606;1;T;T;;0;0
203;2035606;1;T;T;;0;0
204;2035606;1;T;T;;0;0
224;2035606;1;T;T;;0;0
938;127877;935000000;;T;;859814;0
938;624505;4000000000;;T;;1985922;859814
938;859814;373082010;;T;;624505;127877
938;1985922;320000;;T;;3716059;624505
938;3716059;75700000000;;T;;0;1985922
950;107597;1000000000;;T;;107598;0
950;107598;1000000000;;T;;107600;107597
950;107600;1000000000;;T;;107602;107598
950;107602;999997650;;T;;107604;107600
950;107604;1000000000;;T;;107606;107602
950;107606;999997650;;T;;107608;107604
```

and so on and so forth for ~4.8M lines ...

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (manual)

### CODE REVIEW

**Review by tinker-michaelj — APPROVED** [review]
LGTM!

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpTokenAssociationsSubcommand.java [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+    SortedMap<Pair<Long, Long>, TokenRel> gatherTokenAssociations(@NonNull TokenRelStorageAdapter tokenRelStore) {
+        requireNonNull(tokenRelStore, "tokenRelStore");
+
+        final SortedMap<Pair<Long, Long>, TokenRel> r = new TreeMap<>(getPairOfLongsComparator());
+
+        if (tokenRelStore.areOnDisk()) {
+            final var tokenRels = requireNonNull(tokenRelStore.getOnDiskRels());
+            final var tokenAssociations = new ConcurrentLinkedQueue<TokenRel>();
+            try {
+                final var threadCount = 8; // Good enough for my laptop, why not?
```
Comment (mhess-swl) [inline hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpTokenAssociationsSubcommand.java:226]: May be good to make this configurable at some point

**Review by mhess-swl — APPROVED** [review]

**Review by povolev15 — APPROVED** [review]

File: hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpTokenAssociationsSubcommand.java (hunk 1 above)
Comment (povolev15) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpTokenAssociationsSubcommand.java:226]: +1

Comment (david-bakin-sl) [reply hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpTokenAssociationsSubcommand.java:226]: I'll do it in several of these dumpers at the same time, separate PR.

**Pull request merged**